tests/aead.cc

```diff
@@ -0,0 +1,171 @@
+                                                            aad.data(),
+                                                            aad.size(),
+                                                            msg.data(),
+                                                            msg.size(),
+                                                            got_cipher.data(),
+                                                            got_tag.data());
+
+    if (valid) {
+      ASSERT_EQ(res, 0);
+      ASSERT_EQ(cipher, got_cipher);
+      ASSERT_EQ(tag, got_tag);
+    }
+  }
+
+  // Decrypt
+  {
+    bytes got_msg(msg.size());
+    EverCrypt_Error_error_code res = EverCrypt_AEAD_decrypt(state,
+                                                            iv.data(),
+                                                            iv.size(),
+                                                            aad.data(),
+                                                            aad.size(),
+                                                            cipher.data(),
+                                                            cipher.size(),
+                                                            tag.data(),
+                                                            got_msg.data());
+
+    if (valid) {
+      ASSERT_EQ(res, 0);
+      ASSERT_EQ(msg, got_msg);
+    } else {
+      ASSERT_NE(res, 0);
+    }
+  }
+}
+
+// ----- ChaCha20Poly1305 ------------------------------------------------------
+
+typedef EverCryptSuite<WycheproofChacha20Poly1305> ChaCha20Suite;
+
+TEST_P(ChaCha20Suite, KAT)
+{
+  EverCryptConfig config;
+  WycheproofChacha20Poly1305 test;
+  tie(config, test) = this->GetParam();
+
+  EverCrypt_AEAD_state_s* state;
+  EverCrypt_AEAD_create_in(
+    Spec_Agile_AEAD_CHACHA20_POLY1305, &state, test.key.data());
+
+  encrypt_decrypt(
+    state, test.iv, test.aad, test.msg, test.ct, test.tag, test.valid);
+
+  EverCrypt_AEAD_free(state);
+}
+
+// ----- AES GCM -------------------------------------------------------------
+
+typedef EverCryptSuite<WycheproofAesGcm> AesGcmSuite;
+
+TEST_P(AesGcmSuite, KAT)
+{
+  EverCryptConfig config;
+  WycheproofAesGcm test;
+  tie(config, test) = this->GetParam();
+
+  EverCrypt_AEAD_state_s* state;
+
+  if (test.iv.size() != 12) {
+    cout << "Skipping ivSize != 96" << endl;
+    return;
+  }
+
+  EverCrypt_Error_error_code res;
+  if (test.keySize == 128) {
+    res = EverCrypt_AEAD_create_in(
+      Spec_Agile_AEAD_AES128_GCM, &state, test.key.data());
+  } else if (test.keySize == 192) {
+    cout << "Skipping keySize = 192" << endl;
+    return;
+  } else if (test.keySize == 256) {
+    res = EverCrypt_AEAD_create_in(
+      Spec_Agile_AEAD_AES256_GCM, &state, test.key.data());
+  } else {
+    FAIL() << "Unexpected keySize.";
+  }
+
+  if (res != EverCrypt_Error_Success) {
+    cout << "Skipping test due to EverCrypt error" << res << endl;
```

Comment:
Now we don't know whether this was to expect or not.
We have the information in the config that we could use to figure out whether this is supposed to fail or not. But for some platforms this should always fail so I'm not sure what the best way is to test this other than not running it on platforms where we know that it's not supported.

Reply:
Good point. I implemented this for AES-CCM, too, and only noted later that no test was ever run due to this error being silent. I think we should definitely address this. As a first step I could check if something (aesni, ...) was disabled. If not, it must work in CI. Not sure about other platforms.